### ISSUE TITLE
Type Index

### ISSUE THREAD

**Issue by ShishkinDmitriy** [npgall/cqengine#100]
Is any plans to introduce some kind of TypeIndex? To retrieve subclasses of holding elements, for example when ```Collection<? extends T>```. In that case I want to have indexes of some subclasses of ```T```

**Comment by npgall** [comment]
There is an example of how to use inheritance with CQEngine [here](https://github.com/npgall/cqengine/tree/master/code/src/test/java/com/googlecode/cqengine/examples/inheritance). Does this help?

**Comment by ShishkinDmitriy** [comment]
Thanks!
What about subclasses?
```
Query<Car> cars = equal(CAR_TYPE, Car.class);
```
Result set will not contain SportCars?

**Comment by npgall** [comment]
The ResultSet can contain a mixture of Car and SportsCar (or any other subtype of) objects.

You just need to use `instanceof` to check the type, and cast to the appropriate subtype while you're iterating the results, if you need to access members which are only contained in a particular subclass.

**Comment by ShishkinDmitriy** [comment]
You mean that I should iterate over all and then filter ResultSet by type.
But I have a collection with disjoint elements.
```
abstract class Base {}
class A extends Base {}
class B extends Base {}
...
Collection<Base> all = new HashSet<>();
Query<A> a= equal(TYPE, A.class);  // will iterate all elements
Query<B> a= equal(TYPE, B.class);  // will iterate all elements
```
When I have interfaces it gets worse - class can get multiple interfaces.
For example, you have ```Collection<Object>``` and you want to retrieve all ```and(instanceof(Serializable.class), instanceof(Clonable.class))```.
This example is a bit synthetic but reflect the main idea.

I think it should be one more index, because ```instanceof``` behave not like ```equals```.

**Comment by ShishkinDmitriy** [comment]
I will try to give more clear example with disjount classes.
```
class GraphObject {}
class Node extends GraphObject {}
class Link extends GraphObject {}
class Graph implements Collection<GraphObject> {}
```
I understand that I can implement ```Graph``` as 2 collections - ```Collection<Node>``` and ```Collection<Link>```. 
But there is some existing interface with ```<T> T doSome(Collection<? extends T>)```.
And I have to wrap that 2 collection into 1 - ```Collection<GraphObject>```.
This case is similar to #87.

**Comment by npgall** [comment]
Hi Shishkin,

No I mean it works both ways.

- If you are looking for a particular subtype of objects, then you can compose a query which requires that results match a subtype-specific attribute, and then only objects of that particular subtype will be returned.

- On the other hand, if you are looking for objects which may be a mix of subtypes, then you can compose a query which only refers to their shared attributes, and then a mix of object types will be returned.

The indexes can be disjoint, if you build them on _nullable_ attributes.

In the inheritance example, note that the subclass SportsCar has a 'horsepower' field, but the superclass Car does not. The SportsCar has a nullable attribute called [HORSEPOWER](https://github.com/npgall/cqengine/blob/master/code/src/test/java/com/googlecode/cqengine/examples/inheritance/SportsCar.java#L42) whose generic type is Car, but which is programmed to only return a horsepower value if the Car is a SportsCar. So if you built an index on the HORSEPOWER attribute, then that index would only contain SportsCars.

Does this help?

**Comment by ShishkinDmitriy** [comment]
I need all subtree, not only particular subclass.
Let's say I have hierarchy as below.
![image1](https://cloud.githubusercontent.com/assets/2331042/20216771/7bd00cee-a82d-11e6-83a9-d380d0f84c73.gif)
To query all ```Mammals``` from ```Collection<Animal>``` I have to 

1. Add Attribute ```Attribute<Animal, Class<? extends Animal>> TYPE  = ... { return animal.getClass(); }```
2. Write query 
```
or(
   equals(TYPE, Mammal.class), 
   equals(TYPE, Bat.class), 
   equals(TYPE, Whale.class), 
   equals(TYPE, SeaLion.class)
)
```

Is it right solution? What if I have a complex and wide hierarchy?

**Comment by npgall** [comment]
That's a cool example! Yes that would work.

I understand your concern about the query becoming large for wide hierarchies. But this is somewhat of a necessity if you want to be able to index your collection by the type of your objects. The following are alternative options - some indexable, some not.

**Helper function + in() query on regular attribute (indexable)**
To make your query above more concise, you could alternatively use an in() query, combined with a function in your application which returned the subclasses of a given supertype. Of course, CQEngine has no idea of how the inheritance hierarchy in your application is structured, so it's up to you how to implement such a function. You might need some mechanism to register all possible types at startup, or similar. Regardless, this approach will leverage an index on the TYPE attribute.

```java
static Collection<Class<? extends Animal>> subtypesOf(Class<? extends Animal> supertype) {
    // return all classes which are subtypes of the given supertype
}
...
Query<Animal> mammals = in(TYPE, subtypesOf(Mammal.class));
```

**Query DSL (indexable)**
Here's a way to (optionally) define your own query DSL to make the query above slightly more readable.
```java
public static <O> Query<O> isTypeOf(Class<?> type) {
    return in(TYPE, subtypesOf(type));
}
...
Query<Animal> mammals = isTypeOf(Mammal.class);
```

Alternatively, you can also leverage virtual attributes...

**Static virtual attributes (indexable)**
The following is an example of a static virtual attribute. Unlike a regular attribute which just reads a value from a field, a _virtual_ attribute encapsulates some kind of function to compute the value. 

The caveat with this approach is you'd need to define a virtual attribute for each supertype you want to query. Nonetheless, you can build an index on this static virtual attribute to accelerate retrieval of objects which are mammals.
```java
static final Attribute<Animal, Boolean> IS_MAMMAL = attribute(a -> a instanceof Mammal);
...
Query<Animal> mammals = equal(IS_MAMMAL, true);
```

**Dynamic virtual attributes (not indexable)**
To avoid having to define multiple static attributes up-front, you can also define attributes dynamically - inline in queries. However the tradeoff is you can't build indexes on attributes created this way.
```java
Query<Animal> mammals = equal(attribute(a -> a instanceof Mammal), true);
```

**Query DSL based on dynamic virtual attributes (not indexable)**
Here's a way to wrap the approach with dynamic attributes, to achieve the same syntax as in the DSL we had above which was based on static attributes. As this is still based on dynamic attributes, it can't be indexed.
```java
public static <O> Query<O> isTypeOf(Class<?> type) {
    return equal(attribute(a -> type.isAssignableFrom(a.getClass())), true);
}
...
Query<Animal> mammals = isTypeOf(Mammal.class);
```
So as you can see there are a few ways to do this, each with pros and cons!

**Comment by ShishkinDmitriy** [comment]
Thanks, it is good explanation.
I am interesting now only in indexable approaches.
Initially, when I created issue I meant that it is needed index (mapping) ```Class -> Animal```. With arguments you can hold mapping ```Boolean -> Animal```, where ```Boolean``` value is whether an instance of ```Class``` or not. 
I can agreee with than solution.
But I still confused a bit.
```
Collection      |  0 1 2 3 4 5 6 7 8 9    // hold 10 animals
------------------------------------------------------------
Index(Mammal)   |  + + + + + + - - - -    // true -> {0,1,2,3,4,5}, false -> {6,7,8,9}  // size=10
Index(Bat)      |  + + - - - - - - - -    // true -> {0,1}, false -> {2,3,4,5,6,7,8,9}  // size=10
Index(Whale)    |  - - + + - - - - - -    // true -> {2,3}, false -> {0,1,4,5,6,7,8,9}  // size=10
Index(SeaLion)  |  - - - - + + - - - -    // true -> {4,5}, false -> {0,1,2,3,6,7,8,9}  // size=10
------------------------------------------------------------
Partial(Mammal) |  + + + + + +            // true -> {0,1,2,3,4,5}  // size=6
Partial(Bat)    |  + +                    // true -> {0,1} // size=2
Partial(Whale)  |      + +                // true -> {2,3} // size=2
Partial(SeaLion)|          + +            // true -> {4,5} // size=2
```
Can I construct Partial Index from this indexes? To hold only ```true``` values. 
And ```false``` values retrieve from ```not(equal(IS_MAMMAL, true))```

**Comment by npgall** [comment]
Actually I don't think you need to worry about that.

A Boolean index can actually hold 4 states for each object: true, false, {no value}, {multiple values}. The set of states which are allowed is determined by the type of attribute you build the index on. The nullable and multi-value attributes allow you to build _sparse indexes_.

For example, if you build a HashIndex on a:
* SimpleAttribute, then the index is guaranteed to contain exactly one entry for each object: true or false
* SimpleNullableAttribute, then the index can contain _at most_ one entry for each object: true, false, or {no entry}
* MultiValueAttribute or MultiValueNullableAttribute, then the index can contain zero or more entries for each object: true, false, {no entry}, {multiple entries}. 

The multi-value attributes are not really intended for use with boolean attributes (usually it makes no sense for an object to have both a true and a false value at the same time). However they can be useful for indexing objects where each object can have multiple values like keywords associated with it.

Basically, if you use a SimpleNullableAttribute which returns null instead of false, then the index will only contain entries for the objects which are 'true'.

**Comment by ShishkinDmitriy** [comment]
Summing up, if I will:

1. create ```new SimpleNullableAttribute(a -> (a instanceof Mammal) ? true : null)```;
2. create index base on it;
3. add a lot of oviparous objects into collection (Crocodiles, Geese and Pelicanes);

then the Index size **will NOT** grow?

**Comment by npgall** [comment]
Yes that's correct.

On 15 Nov 2016 13:21, "Dmitriy Shishkin" <notifications@github.com> wrote:

Summing up, if I will:

   1. create new SimpleNullableAttribute(a -> (a instanceof Mammal) ? true
   : null);
   2. create index base on it;
   3. add a lot of oviparous objects into collection (Crocodiles, Geese and
   Pelicanes);

then the Index size *will NOT* grow?

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/npgall/cqengine/issues/100#issuecomment-260639423>, or mute
the thread

**Comment by npgall** [comment]
I'll close this issue now then. Feel free to open another if you need anything!
